### PR TITLE
MAHOUT-900 [IQP] Implement FWT optimization for IQP encoding

### PR DESCRIPTION
**Pull request by ryankert01**
…corresponding tests and benchmarks

- Added Fast Walsh-Hadamard Transform (FWT) optimization to IQP encoding, reducing complexity from O(4^n) to O(n * 2^n).
- Introduced new kernels for phase computation and FWT stages in both single and batch encoding.
- Updated kernel configuration with thresholds for shared memory usage.
- Created tests to verify normalization and correctness of FWT-optimized IQP states.
- Added benchmarking script to measure performance improvements from FWT optimization.

### Changes Made

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring
- [ ] Documentation
- [ ] Test
- [ ] CI/CD pipeline
- [ ] Other

### Breaking Changes

- [ ] Yes
- [x] No

### Checklist

- [ ] Added or updated unit tests for all changes
- [ ] Added or updated documentation for all changes
- [ ] Successfully built and ran all unit tests or manual tests locally
- [ ] PR title follows "MAHOUT-XXX: Brief Description" format (if related to an issue)
- [ ] Code follows ASF guidelines

### CODE REVIEW

**Comment by ryankert01** [comment]
I will solve the pre-commit and delete single purpose benchmark script later.

File: testing/qdp/test_bindings.py [hunk 1]
```diff
@@ -1080,3 +1024,190 @@ def test_iqp_encode_errors():
...
+    import torch
+    from _qdp import QdpEngine
+
+    if not torch.cuda.is_available():
+        pytest.skip("GPU required for QdpEngine")
+
+    engine = QdpEngine(0)
+
+    # Test at and around the shared memory threshold
+    for num_qubits in [9, 10]:
```
Comment (guan404ming) [inline testing/qdp/test_bindings.py]: Could we add n=11 so it actually straddles the boundary and exercises the global-memory multi-launch?

File: qdp/qdp-kernels/src/iqp.cu
Comment (guan404ming) [inline qdp/qdp-kernels/src/iqp.cu:1]: Would it be worth adding an early `cudaGetLastError()` check right after the phase kernel to fail fast?

**Comment by guan404ming** [comment]
Please help resolve the ci error as well, thanks!

**Comment by guan404ming** [comment]
Small nit: we could make the pr title more short and concise. Current one is a little bit too long.

**Comment by ryankert01** [comment]
I will upload benchmark result shortly, Thanks! (then, I will delete the single purpose benchmark!)
cc @rich7420

**Comment by guan404ming** [comment]
I think we should only merge the green ci pr. Please let me know if there is any place need help, thanks!

**Comment by ryankert01** [comment]
@guan404ming Don't worry, I don't address them now only because I'm on my phone. Will address all the issues shortly.

**Comment by guan404ming** [comment]
Hahaha, take your time. I'm chilling at Cafe shop.

File: testing/qdp/test_bindings.py (hunk 1 above)
Comment (ryankert01) [reply testing/qdp/test_bindings.py]: Nice catch, updated!

**Comment by ryankert01** [comment]
Comparison of Naive O(4^n) vs Fast Walsh-Hadamard Transform O(n×2^n) implementations.

## Single Sample Performance

| Qubits | Naive | FWT | Speedup |
|--------|-------|-----|---------|
| 4 | 0.150 ms | 0.150 ms | 1.0x |
| 6 | 0.237 ms | 0.143 ms | 1.7x |
| 8 | 1.199 ms | 0.143 ms | 8.4x |
| 10 | 6.213 ms | 0.117 ms | **53x** |

## Batch Performance (64 samples)

| Qubits | Naive | FWT | Speedup |
|--------|-------|-----|---------|
| 4 | 0.139 ms | 0.149 ms | 0.9x |
| 6 | 0.316 ms | 0.155 ms | 2.0x |
| 8 | 1.223 ms | 0.139 ms | 8.8x |
| 10 | 22.23 ms | 0.165 ms | **135x** |

Thanks for the review!

File: qdp/qdp-kernels/src/iqp.cu
Comment (ryankert01) [reply qdp/qdp-kernels/src/iqp.cu:1]: I chose to not do it because it only save 0.1 ms (see benchmark results)

**Comment by rich7420** [comment]
@ryankert01 
sorry for the late.
You're doing well.
thanks for the patch!
